Ticket 3 send images

nigel-f-wilson opened the pull request:
## Description
This PR addresses this issue [ticket 3: sending images](https://github.com/nolastemgarden/445d6c/issues/4).
The assignment asked that I explain a couple of different ways we could have handled the UX for attaching an image.  See my notes in the next section.

## Notes on your approach and thought process
At first, I began building a completely custom Dialog using Material UI components. When I added and `<input type="file">` to the dialog I realized that that was going to open an OS-native dialog for actually selecting files from the device's file storage. I did not like the UX of a button that opened a dialog that opened a second dialog so I decided to get rid of the custom dialog I had built and simply add an icon button to the existing `<form>`  that opened the native file picker dialog.  
So far I have only tested this app on Windows.  

I could have handled the previewing of selected photos differently but I am fond of the way I have it because it makes it clear to the user when a photo is part of a sent message vs. when it is only a preview, loaded to the browser but not yet sent to Cloudinary.

It is also worth noting that Cloudinary offers its own Photo Selector Widget for React that looks like it would be pretty easy to implement, but I chose not to use it because it offers features more advanced than the scope of this issue such as accessing the device camera and taking new pictures.

nigel-f-wilson committed:
Create ImagePickerDialog.js

nigel-f-wilson committed:
Add IconButton to open send image dialog

nigel-f-wilson committed:
Update createMuiTheme to createTheme

nigel-f-wilson committed:
Build Dialog with Image preview and button to open OS file picker

nigel-f-wilson committed:
Add state for image files

nigel-f-wilson committed:
Merge branch 'main' into ticket-3-send-images

nigel-f-wilson committed:
Add ref to make MUI button control html input

nigel-f-wilson committed:
Reduce layout height

The logout button was pushing the user input field off the page.

nigel-f-wilson committed:
Delete ImagePickerDialog.js

No need for custom build MUI dialog

nigel-f-wilson committed:
Remove redundant overflow scroll

nigel-f-wilson committed:
Rename SenderBubble and OtherUserBubble

These are no longer the actual bubbles but wrappers for various message layouts depending on number of attachments.

nigel-f-wilson committed:
Add image uploading to form submit and pic preview to input

nigel-f-wilson committed:
Create Message.js

A versatile component that can render a message for either user and with any number of attachments

nigel-f-wilson committed:
Rename SenderBubble and OtherUserBubble

These are no longer the actual bubbles but wrappers for various message layouts depending on number of attachments.

FilipeGorodscy reviewed (commented):
Thank you for your work on this PR.

Let us know when the changes are implemented by using the "/hatchy pr" command on Slack again and another reviewer will come by to give this PR another look and give you approval 👍🏻

nigel-f-wilson committed:
Update client/src/components/ActiveChat/Message.js

use optional chaining to determine display setting of picture in text bubble based on number of attachments.

Co-authored-by: FilipeGorodscy <53091098+FilipeGorodscy@users.noreply.github.com>

nigel-f-wilson committed:
Create .env

To store Cloudinary upload keys

nigel-f-wilson committed:
Switch to correct icon

nigel-f-wilson committed:
Replace Cloudinary URLs with values imported from process.env

nigel-f-wilson committed:
Standardize syntax

Declared function using const and an arrow function rather than function keyword.

nigel-f-wilson committed:
Adjust picture sizing

nigel-f-wilson committed:
Clean up syntax with optional chaining

attachments?.length === 1 works to BOTH verify attachments is not null and not longer than 1.

nigel-f-wilson committed:
Merge branch 'ticket-3-send-images' of https://github.com/nolastemgarden/445d6c into ticket-3-send-images

nigel-f-wilson committed:
Update MUI icon library to v5

nigel-f-wilson committed:
Make message use props based styling, not inline

nigel-f-wilson committed:
Extract TimeString to own functional component

nigel-f-wilson committed:
Extract PictureRow to own functional component

nigel-f-wilson committed:
Update Message.js

Pass down "type" prop rather than "bubbleClassName"

nigel-f-wilson committed:
Extract uploadFilesToCloudinary to thunkCreators

nigel-f-wilson committed:
Add error catching to uploadFilesToCloudinary()

nigel-f-wilson committed:
Switch fetch to axios.post

nigel-f-wilson committed:
Switch for loop to Promise.all

nigel-f-wilson committed:
Remove extraneous console.log

nigel-f-wilson committed:
Update thunkCreators.js

Try using Axios instead of Fetch

FilipeGorodscy reviewed (approved):
Thank you for your work on this PR.

Please merge this PR once you've addressed all comments remaining. 👍

nigel-f-wilson committed:
Use axios transform request

Co-authored-by: FilipeGorodscy <53091098+FilipeGorodscy@users.noreply.github.com>

nolastemgarden commented on client/src/store/utils/thunkCreators.js:
```diff
@@ -79,7 +79,11 @@ export const fetchConversations = () => async (dispatch) => {
 
 const saveMessage = async (body) => {
-  const { data } = await axios.post("/api/messages", body);
+  const { data } = await axios({
+    method: "post", 
+    url: "/api/messages", 
+    data: body
+  });
```
Reverted to `axios.post()`

nigel-f-wilson committed:
Revert to using `axios.post`

nigel-f-wilson committed:
Merge branch 'ticket-3-send-images' of https://github.com/nolastemgarden/445d6c into ticket-3-send-images

nigel-f-wilson committed:
Add missing `await`

nigel-f-wilson committed:
Update thunkCreators.js

Use axios.post instead of fetch.

nigel-f-wilson committed:
Use prop-based styling

Several components had different styling depending on type sender/recipient. Now handled in makeStyles rather than in each component.
